jitstats.lua

```diff
@@ -0,0 +1,462 @@
+
+for i = 0, #vmdef.bcnames-1, 6 do
+    local name = string.sub(vmdef.bcnames, i, i+6):match "^%s*(.-)%s*$"
+    bcnames[bccount] = name
+    bcname_lookup[name] = bccount
+    bccount = bccount + 1
+end
+
+local traceerror_info = {
+  {name = "RECERR",  msg = "error thrown or hook called during recording"},
+  {name = "TRACEUV", msg = "trace too short"},
+  {name = "TRACEOV", msg = "trace too long"},
+  {name = "STACKOV", msg = "trace too deep"},
+  {name = "SNAPOV",  msg = "too many snapshots"},
+  {name = "BLACKL",  msg = "blacklisted"},
+  {name = "RETRY",   msg = "retry recording"},
+  {name = "NYIBC",   msg = "NYI: bytecode %d"},
+  
+  -- Recording loop ops.
+  {name = "LLEAVE",  msg = "leaving loop in root trace"},
+  {name = "LINNER",  msg = "inner loop in root trace"},
+  {name = "LUNROLL", msg = "loop unroll limit reached"},
+  
+  -- Recording calls/returns.
+  {name = "BADTYPE", msg = "bad argument type"},
+  {name = "CJITOFF", msg = "JIT compilation disabled for function"},
+  {name = "CUNROLL", msg = "call unroll limit reached"},
+  {name = "DOWNREC", msg = "down-recursion, restarting"},
+  {name = "NYIFFU",  msg = "NYI: unsupported variant of FastFunc %s"},
+  {name = "NYIRETL", msg = "NYI: return to lower frame"},
+  
+  -- Recording indexed load/store.
+  {name = "STORENN", msg = "store with nil or NaN key"},
+  {name = "NOMM",    msg = "missing metamethod"},
+  {name = "IDXLOOP", msg = "looping index lookup"},
+  {name = "NYITMIX", msg = "NYI: mixed sparse/dense table"},
+  
+  -- Recording C data operations. */
+  {name = "NOCACHE", msg = "symbol not in cache"},
+  {name = "NYICONV", msg = "NYI: unsupported C type conversion"},
+  {name = "NYICALL", msg = "NYI: unsupported C function type"},
+  
+  -- Optimizations.
+  {name = "GFAIL",   msg = "guard would always fail"},
+  {name = "PHIOV",   msg = "too many PHIs"},
+  {name = "TYPEINS", msg = "persistent type instability"},
+  
+  -- Assembler.
+  {name = "MCODEAL", msg = "failed to allocate mcode memory"},
+  {name = "MCODEOV", msg = "machine code too long"},
+  {name = "MCODELM", msg = "hit mcode limit (retrying)"},
+  {name = "SPILLOV", msg = "too many spill slots"},
+  {name = "BADRA",   msg = "inconsistent register allocation"},
+  {name = "NYIIR",   msg = "NYI: cannot assemble IR instruction %d"},
+  {name = "NYIPHI",  msg = "NYI: PHI shuffling too complex"},
+  {name = "NYICOAL", msg = "NYI: register coalescing too complex"},
+}
+
+local traceerr, traceerr_fromid = {}, table.new(#vmdef.traceerr, 0)
+
+for i, evt in pairs(traceerror_info) do
+    local id = traceerr_lookup[evt.msg]
+
+    if id then
+        traceerr[evt.name] = id
+        traceerr_fromid[id] = evt.name
+    end
+end
+
+local TRERR_BLACKL, TRERR_LLEAVE, TRERR_LUNROLL, TRERR_NYIBC = traceerr.BLACKL, traceerr.LLEAVE, traceerr.LUNROLL, traceerr.NYIBC
+
+local function printf(fmt, ...)
+    if select("#", ...) == 0 then
+        print(fmt)
+    else
+        print(string.format(fmt, ...))
+    end
+end
```

Comment:
no need for the condition; `string.format(s)` is valid.  my typical `printf()` looks like
```
local function printf(fmt, ...)
    io.write(fmt:format(...))
end
```
(note it doesn't add a newline, just like in C)
but can be as simple as:
```
local function printf(...)
   print(string.format(...))
end
```

Reply:
I was probably using it with non sting values in the past but i was also thinking of letting the user supply there own output function as well

Reply:
fixed